src/vsr/journal.zig

```diff
@@ -703,83 +843,505 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
+        ///
+        ///
+        /// When replica_count=1, repairing broken/lost prepares over VSR is not an option, so
+        /// if a message is faulty the replica will abort.
+        ///
+        ///
+        /// Recovery decision table:
+        ///
+        ///   label                   @A  @B  @C  @D  @E  @F  @G  @H  @I  @J  @K  @L
+        ///   header valid             0   1   1   0   0   1   1   1   1   1   1   1
+        ///   header reserved          _   1   0   _   _   1   0   1   0   0   0   0
+        ///   prepare valid            0   0   0   1   1   1   1   1   1   1   1   1
+        ///   prepare reserved         _   _   _   1   0   0   1   1   0   0   0   0
+        ///   match checksum           _   _   _   _   _   _   _  !1   0   0   0   1
+        ///   match op                 _   _   _   _   _   _   _  !1   <   >   1  !1
+        ///   match view               _   _   _   _   _   _   _  !1   _   _  !0  !1
+        ///   decision (replicas>1)  vsr vsr vsr vsr vsr vsr vsr nil fix vsr vsr eql
+        ///   decision (replicas=1)                  fix fix
+        ///
+        /// Legend:
+        ///
+        ///    0  false
+        ///    1  true
+        ///   !0  assert false
+        ///   !1  assert true
+        ///    _  ignore
+        ///    <  header.op < prepare.op
+        ///    >  header.op > prepare.op
+        ///
+        /// A "valid" header/prepare:
+        /// 1. has a correct checksum, AND
+        /// 2. one of the following:
+        ///   * command=reserved
+        ///   * command=prepare AND correct cluster AND is in its proper slot (op % slot_count).
+        ///
+        ///
+        /// Regarding cases @E and @F for *single-replica clusters*:
+        ///
+        /// Most likely the redundant header write was corrupted/lost (respectively).
+        /// Since single-replica clusters never rewrite ops, it can be locally repaired by copying
+        /// over the prepare's header.
+        ///
+        /// However, another (astronomically low) possibility is this:
+        ///
+        /// 1. A prepare is written to slot 12.
+        /// 2. A redundant header is written to slot 12.
+        /// 3. Continue normal operation; wrap around the WAL.
+        /// 4. A prepare is written to slot 12, but lost.
+        /// 5. A redundant header is written to slot 12.
+        /// 6. Replica crash.
+        /// 7. On recovery, the old prepare (from step 1) is read and used to "repair" the corrupt
+        ///    redundant header.
+        ///
+        /// This would be a safety violation — we assume that the likelihood of this event is
+        /// negligible.
+        fn recover_prepares_on_read(completion: *Storage.Read) void {
+            const read = @fieldParentPtr(Self.Read, "completion", completion);
+            const self = read.self;
+            const replica = @fieldParentPtr(Replica, "journal", self);
+            assert(!self.recovered);
```

Comment:
```suggestion

            assert(!self.recovered);
```

To group the assertion section, also since it's not directly related to the replica constant assignment.